test/distributed/_composable/test_composability/test_3d_composability.py

```diff
@@ -0,0 +1,565 @@
+                model.train()
+
+        else:
+            if parallel_dims.tp_enabled:
+                self.apply_tp(
+                    model,
+                    world_mesh["tp"],
+                    loss_parallel=parallel_dims.loss_parallel_enabled,
+                    enable_async_tp=False,
+                )
+
+            if parallel_dims.dp_shard_enabled:
+                dp_mesh_dim_names = (
+                    ("dp_replicate", "dp_shard")
+                    if parallel_dims.dp_replicate_enabled
+                    else ("dp",)
+                )
+                dp_mesh = world_mesh[dp_mesh_dim_names]
+                self.apply_fsdp(
+                    model,
+                    world_mesh["dp"],
+                    param_dtype=torch.bfloat16,
+                    reduce_dtype=torch.float32,
+                    tp_enabled=parallel_dims.tp_enabled,
+                    pp_enabled=parallel_dims.pp_enabled,
+                    cpu_offload=None,
+                )
+            model.to_empty(device=device)
+            model.train()
+            models = [model]
+
+        optimizer_kwargs = {
+            "lr": 0.01,
+            "betas": (0.9, 0.95),
+            "weight_decay": 0.1,
+            "fused": False,
+            "foreach": True,
+        }
+        optimizers = [
+            torch.optim.AdamW(model.parameters(), **optimizer_kwargs)
+            for model in models
+        ]
+
+        for train_step in range(10):
+            for optimizer in optimizers:
+                optimizer.zero_grad()
+
+            # Generate random data
+            data = torch.rand(4, 8, device=device)
+            labels = torch.randint(0, 2, (4,), device=device)
+            if parallel_dims.pp_enabled:
+                is_last_stage = pp_mesh.get_local_rank() == pp_mesh.size() - 1
+                if pp_mesh.get_local_rank() == 0:
+                    pp_schedule.step(data)
+                elif is_last_stage:
+                    losses = []
+                    pp_schedule.step(target=labels, losses=losses)
+                else:
+                    pp_schedule.step()
+
+                # accumulate losses across pipeline microbatches
+                loss = (
+                    torch.mean(torch.stack(losses))
+                    if is_last_stage
+                    else torch.Tensor([-1.0])
+                )
+            else:
+                # Forward
+                output = model(data)
+                # Compute loss
+                loss = loss_fn(output, labels)
+                # Backward
+                loss.backward()
+
+            self.clip_grad_norm_(
+                [param for model in models for param in model.parameters()],
+                1.0,
+                foreach=True,
+                pp_mesh=pp_mesh if parallel_dims.pp_enabled else None,
+            )
+
+            for optimizer in optimizers:
+                optimizer.step()
+
+    def pipeline_llama_manual_split(
```

Comment:
suggest you delete the 2 helpers 'pipeline_llama_manual_split' and 'generate_split_points'.

1. it is complicated code to read, and it is complicated _because_ it is supposed to work on a real llama model and be flexible to cmdline for different split points
2. we do not need flexibility in the unit test and we are not using a llama model.  There is no `.tok_embeddings` field in the model
3. it would be useful to test the helper functions _in_ torchtitan, but we already do that.  copying them and then testing the copy is not useful in and of itself.

Conclusion- better to just construct N MLPModel()  objects and assign them to PipelineStage classes yourself.  It should be easy to do in 5 lines of code instead of 90